File: sdks/java/fireweave-sdk/src/main/java/ai/fireweave/sdk/Capabilities.java
package ai.fireweave.sdk;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Capability matrix snapshot (spec {@code capabilities.schema.json}) returned by
 * {@code FireweaveClient.capabilities().get()}. Immutable.
 */
public final class Capabilities {

    public static final String LANGUAGE = "java";
    public static final String SPEC_VERSION = "0.1.0";
    public static final String SDK_VERSION = "0.1.0-SNAPSHOT";
    public static final String PROVIDER_NAME = "fireweave";
    public static final String OPENFEATURE_SPEC_FLOOR = "0.8.0";
    /** 2^53 - 1 cross-language safe integer bound. */
    public static final long INT_SAFE_MAX_ABS = 9007199254740991L;

    private final String backend;
    private final LifecycleState lifecycle;
    private final Map<String, Boolean> staticFeatures;
    private final Map<String, Boolean> runtimeFeatures;
    private final List<String> capabilityNames;

    Capabilities(String backend,
                 LifecycleState lifecycle,
                 Map<String, Boolean> staticFeatures,
                 Map<String, Boolean> runtimeFeatures,
                 List<String> capabilityNames) {
        this.backend = backend;
        this.lifecycle = lifecycle;
        this.staticFeatures = Collections.unmodifiableMap(new LinkedHashMap<>(staticFeatures));
        this.runtimeFeatures = Collections.unmodifiableMap(new LinkedHashMap<>(runtimeFeatures));
        this.capabilityNames = Collections.unmodifiableList(new ArrayList<>(capabilityNames));
    }

    public String backend() {
        return backend;
    }

    public LifecycleState lifecycle() {
        return lifecycle;
    }

    public Map<String, Boolean> staticFeatures() {
        return staticFeatures;
    }

    public Map<String, Boolean> runtimeFeatures() {
        return runtimeFeatures;
    }

    /** Negotiated dotted capability names, e.g. "releases.setContext" (dynamic dispatch sugar). */
    public List<String> names() {
        return capabilityNames;
    }

    public boolean supports(String capabilityName) {
        return capabilityNames.contains(capabilityName);
    }

    /**
     * Canonical structured {@code {static, runtime}} capability matrix per
     * {@code spec/capabilities.schema.json} (orchestrator ruling 18: capabilities.get is the
     * structured matrix, never a flat capability-string list).
     */
    public JsonValue toJsonValue() {
        Map<String, JsonValue> openFeature = new LinkedHashMap<>();
        openFeature.put("specFloor", JsonValue.of(OPENFEATURE_SPEC_FLOOR));
        openFeature.put("providerName", JsonValue.of(PROVIDER_NAME));
        openFeature.put("serverOnly", JsonValue.of(true));

        Map<String, JsonValue> staticFeaturesJson = new LinkedHashMap<>();
        staticFeatures.forEach((k, v) -> staticFeaturesJson.put(k, JsonValue.of(v)));

        Map<String, JsonValue> staticNode = new LinkedHashMap<>();
        staticNode.put("language", JsonValue.of(LANGUAGE));
        staticNode.put("sdkVersion", JsonValue.of(SDK_VERSION));
        staticNode.put("specVersion", JsonValue.of(SPEC_VERSION));
        staticNode.put("openFeature", JsonValue.ofObject(openFeature));
        staticNode.put("features", JsonValue.ofObject(staticFeaturesJson));

        Map<String, JsonValue> runtimeFeaturesJson = new LinkedHashMap<>();
        runtimeFeatures.forEach((k, v) -> runtimeFeaturesJson.put(k, JsonValue.of(v)));

        Map<String, JsonValue> limits = new LinkedHashMap<>();
        limits.put("intSafeMaxAbs", JsonValue.of(INT_SAFE_MAX_ABS));
        limits.put("shutdownTimeoutMsDefault", JsonValue.of(FireweaveConfig.DEFAULT_SHUTDOWN_TIMEOUT_MS));

        Map<String, JsonValue> runtimeNode = new LinkedHashMap<>();
        runtimeNode.put("backend", JsonValue.of(backend));
        runtimeNode.put("lifecycle", JsonValue.of(lifecycle.name()));
        runtimeNode.put("features", JsonValue.ofObject(runtimeFeaturesJson));
        runtimeNode.put("limits", JsonValue.ofObject(limits));

        Map<String, JsonValue> root = new LinkedHashMap<>();
        root.put("static", JsonValue.ofObject(staticNode));
        root.put("runtime", JsonValue.ofObject(runtimeNode));
        return JsonValue.ofObject(root);
    }
}

File: sdks/go/fireweave/capabilities.go
package fireweave

// Version is the Fireweave Go SDK package version.
const Version = "0.1.0"

// SpecVersion is the Fireweave SDK spec version this package implements
// (spec/version.json).
const SpecVersion = "0.1.0"

// openFeatureSpecFloor is the minimum OpenFeature spec version supported.
const openFeatureSpecFloor = "0.8.0"

// intSafeMaxAbs is the cross-language safe integer bound (2^53−1),
// spec/capabilities.schema.json.
const intSafeMaxAbs = int64(9007199254740991)

// Capabilities is the structured capability matrix returned by
// Client.Capabilities().Get(), per spec/capabilities.schema.json
// (orchestrator ruling 18). JSON field names match the schema exactly.
type Capabilities struct {
	// Static describes capabilities always true for this package build.
	Static StaticCapabilities `json:"static"`
	// Runtime describes capabilities that depend on adapter selection and
	// configuration at runtime.
	Runtime RuntimeCapabilities `json:"runtime"`
}

// StaticCapabilities are compile-time capabilities of this language binding.
type StaticCapabilities struct {
	Language    string                  `json:"language"`
	SDKVersion  string                  `json:"sdkVersion"`
	SpecVersion string                  `json:"specVersion"`
	OpenFeature OpenFeatureCapabilities `json:"openFeature"`
	Features    map[string]bool         `json:"features"`
}

// OpenFeatureCapabilities describe the OpenFeature integration.
type OpenFeatureCapabilities struct {
	SpecFloor    string `json:"specFloor"`
	ProviderName string `json:"providerName"`
	ServerOnly   bool   `json:"serverOnly"`
}

// RuntimeCapabilities are adapter/config-dependent capabilities.
type RuntimeCapabilities struct {
	// Backend is one of "posthog", "inmemory", "none", "other".
	Backend string `json:"backend"`
	// Lifecycle is the current runtime lifecycle state (UNINITIALIZED,
	// INITIALIZING, READY, STALE, ERROR, FATAL, SHUTDOWN).
	Lifecycle string `json:"lifecycle"`
	// Features are adapter-reported runtime feature toggles
	// (remoteEvaluation, localEvaluation, exposureEmission, ...).
	Features map[string]bool `json:"features,omitempty"`
	// Limits are effective numeric bounds (intSafeMaxAbs, ...).
	Limits map[string]int64 `json:"limits,omitempty"`
}

// AdapterCapabilities is the runtime capability report an adapter provides
// through the optional CapabilityReporter interface.
type AdapterCapabilities struct {
	// Backend names the adapter backend ("posthog", "inmemory", "other").
	Backend string
	// Features are backend-specific runtime feature toggles.
	Features map[string]bool
}

// CapabilityReporter is optionally implemented by adapters that can
// describe their runtime capabilities; Client.Capabilities().Get()
// discovers it via type assertion. Adapters that do not implement it are
// reported with backend "other" and no runtime features.
type CapabilityReporter interface {
	ReportCapabilities() AdapterCapabilities
}
